### ISSUE TITLE
If self.user contains unicode characters then str throws UnicodeEncodeError

### ISSUE THREAD

**Issue by jkolo** [cannatag/ldap3#434]
https://github.com/cannatag/ldap3/blob/897591a4f04b0171deb7d66201ff837531808dcc/ldap3/core/connection.py#L350

If `self.user` contains unicode characters then `str` throws exception:
```
2017-11-15 21:09:48,475 - twisted - 131 - ERROR - - Traceback (most recent call last):
2017-11-15 21:09:48,475 - twisted - 131 - ERROR - -   File "/usr/lib64/python2.7/logging/__init__.py", line 861, in emit
2017-11-15 21:09:48,476 - twisted - 131 - ERROR - -     msg = self.format(record)
2017-11-15 21:09:48,477 - twisted - 131 - ERROR - -   File "/usr/lib64/python2.7/logging/__init__.py", line 734, in format
2017-11-15 21:09:48,477 - twisted - 131 - ERROR - -     return fmt.format(record)
2017-11-15 21:09:48,478 - twisted - 131 - ERROR - -   File "/usr/lib64/python2.7/logging/__init__.py", line 465, in format
2017-11-15 21:09:48,479 - twisted - 131 - ERROR - -     record.message = record.getMessage()
2017-11-15 21:09:48,479 - twisted - 131 - ERROR - -   File "/usr/lib64/python2.7/logging/__init__.py", line 329, in getMessage
2017-11-15 21:09:48,480 - twisted - 131 - ERROR - -     msg = msg % self.args
2017-11-15 21:09:48,480 - twisted - 131 - ERROR - -   File "/srv/synapse/lib/python2.7/site-packages/ldap3/core/connection.py", line 350, in __str__
2017-11-15 21:09:48,481 - twisted - 131 - ERROR - -     'user: ' + str(self.user),
2017-11-15 21:09:48,482 - twisted - 131 - ERROR - - UnicodeEncodeError: 'ascii' codec can't encode character u'\u0142' in position 11: ordinal not in range(128)
2017-11-15 21:09:48,482 - twisted - 131 - ERROR - - Logged from file ldap_auth_provider.py, line 133
```

**Comment by jkolo** [comment]
Possible fix is: `'user:' + str(self.user.encode('ascii', 'replace')),`, but then the Unicode characters are converted to "?".

**Comment by cannatag** [comment]
This is not related to ldap3. If you encode to ascii an unicode string you must deal with the conversion of non-printable characters. Or you must be sure you can encode to a "not-only ascii" character set so that the character get printed. The ldap3 logging feature tries to automatically tries to conver to ascii with the ``backslashreplace`` failback for unicode characters, but if you print the unicode to ascii you must directly deal with the encoding.

**Comment by jkolo** [comment]
I think you misunderstood me. When dealin with unicode strings, ldap3 crashes, as shown in the stacktrace in the first post. It's my fix that makes it dispaly questionmarks instead.

**Comment by cannatag** [comment]
Hi, this problem seem related to python 2 only. I've changed the bind() function to honour the auto_encode parameter (defaults to True) even for the binding name. Can you please check the code in dev?

**Comment by cannatag** [comment]
Released in version 2.4.1
